Fixed #26067 -- Added support for ordering ArrayAgg and StringAgg.

florisdenhengst opened the pull request:
Original PR can be found [here](https://github.com/django/django/pull/6886).

Thanks again for the useful discussion!

I processed all of your comments.

Most notable changes are:
- exposure of ordering expressions via get_source_expressions and set_source_expressions. This required some further changes, e.g. overriding `get_source_fields`. This function is used to determine the return type of the expression so this should not include expressions which are used for ordering (these expressions do not affect return value/type).
- support for multiple orderings
- support for string-based specification of ordering in both directions, e.g. similar to `order_by('field')` and `order_by('-field')`

Happy to hear what you guys think!

florisdenhengst commented:
Regarding JSON(B) aggregates: I think it would be best to treat ordering within these aggreagetes in  a separate issue (as was discussed on the [mailinglist](https://groups.google.com/d/msg/django-developers/YCSDX3GApBM/z80uWkJvEgAJ)).

timgraham commented:
Sorry, this missed the feature freeze for 1.11. Could you please rebase it and target it for 2.0 along with release notes in `docs/releases/2.0.txt`?

florisdenhengst commented:
@timgraham no problem, thanks a lot for the head's up. Should be in a better shape now :)

timgraham commented on docs/ref/contrib/postgres/aggregates.txt:
```diff
@@ -22,16 +22,28 @@ General-purpose aggregation functions
+.. class:: ArrayAgg(expression, distinct=False, ordering=None, **extra)
 
     Returns a list of values, including nulls, concatenated into an array.
+    Ordering within the array is unspecified by default. Has one optional
```
"one" seems incorrect. It's best to avoid specifying a number there so that it doesn't have to be updated as things change.

timgraham commented on tests/postgres_tests/test_aggregates.py:
```diff
@@ -22,21 +22,81 @@ class TestGeneralAggregate(PostgreSQLTestCase):
         values = AggregateTestModel.objects.aggregate(arrayagg=ArrayAgg('char_field'))
+        self.assertEqual(values, {'arrayagg': ['Foo1', 'Foo2', 'Foo4', 'Foo3']})
+
+    def test_array_agg_charfield_descending(self):
```
Perhaps some refactoring using `subTest` as done in https://github.com/django/django/pull/7822 would be better.

timgraham commented on django/contrib/postgres/aggregates/general.py:
```diff
@@ -1,14 +1,64 @@
+    def get_source_fields(self):
+        # do not return fields contributed by the ordering expressions, as these should not be used to determine which
+        # the return type of the expression.
+        return [e._output_field_or_none for
```
Use hanging indent:
```
return [
    e._output_field_or_none
   for ...
]
```

florisdenhengst replied:
Fixed ([ln46](https://github.com/django/django/pull/7604/commits/b7cf7ea48cf0785a482b5ebd04558ed9b68a9eac#diff-b1369d2e165bb1dee9ae6196389e53b0R46))

florisdenhengst commented on django/contrib/postgres/aggregates/general.py:
```diff
@@ -1,14 +1,64 @@
+            ordering = [ordering]
+        ordering = ordering or []
+        ordering = [(OrderBy(F(o[1:]), True) if isinstance(o, str) and o[0] == '-' else o) for o in ordering]
+        super(OrderableAgg, self).__init__(
```
Fixed in rebase b7cf7ea48cf0785a482b5ebd04558ed9b68a9eac ([ln 18](https://github.com/django/django/pull/7604/commits/b7cf7ea48cf0785a482b5ebd04558ed9b68a9eac#diff-b1369d2e165bb1dee9ae6196389e53b0R18), [ln 23](https://github.com/django/django/pull/7604/commits/b7cf7ea48cf0785a482b5ebd04558ed9b68a9eac#diff-b1369d2e165bb1dee9ae6196389e53b0R23), [ln 33](https://github.com/django/django/pull/7604/commits/b7cf7ea48cf0785a482b5ebd04558ed9b68a9eac#diff-b1369d2e165bb1dee9ae6196389e53b0R33)
)

florisdenhengst commented on django/contrib/postgres/aggregates/general.py:
```diff
@@ -1,14 +1,64 @@
+        self.expression, self.ordering = exprs[:ordering_exprs_i], exprs[ordering_exprs_i:]
+
+    def get_source_fields(self):
+        # do not return fields contributed by the ordering expressions, as these should not be used to determine which
```
Fixed ([ln43](https://github.com/django/django/pull/7604/commits/b7cf7ea48cf0785a482b5ebd04558ed9b68a9eac#diff-b1369d2e165bb1dee9ae6196389e53b0R43))

florisdenhengst commented:
@timgraham thanks for your comments, really helpful stuff.

I've made the following changes:
- replaced python2-style `super(MyClass, self)` with python3-style `super()`
- hanging indent for list comprehension
- PEP257 verb style for docstring
- removed docstring on number of optional arguments for `ArrayAgg`
- : -> . ! ;)
- correct `versionadded` in docs
- refactored new tests to use `subTest`.

I could 'parameterize' the tests some more by adding the aggregate field as a parameter. We would end up with a single test function for the ordering. I'd love to hear what you think of such a setup, I personally think it will add unnecessary cognitive burden when reading for developers unfamiliar with the intended behavior.

timgraham commented:
Could you do the test refactoring in a separate PR that precedes this one so there's less code churn?

florisdenhengst commented:
The test refactorings only concern tests that were added in this PR, so I don't think your suggestion applies.

I'd be happy to squash the commits though, they were made separately only for the sake of easier review but including some comments in the initial PR (#6886) in a separate commit.

timgraham commented:
@charettes, do you have an input on the high level implementation? I thought it might be related to the work in https://github.com/django/django/pull/8324.

charettes commented:
I think adding support orderable aggregates for `contrib.postgres` only, as opposed to generic `distinct` support in aggregates in #8324, makes more sense here given we don't have any `django.db.models.aggregates` that could benefit from it yet (e.g. it would be a different thing if we had generic `GROUP_CONCAT` support).

timgraham commented on django/contrib/postgres/aggregates/general.py:
```diff
@@ -1,14 +1,63 @@
+        # expression.
+        return [
+            e._output_field_or_none
+            for e in self.get_source_expressions()[:self._get_ordering_expressions_index()]]
```
Put the close ] on the next line.

timgraham commented on django/contrib/postgres/aggregates/general.py:
```diff
@@ -1,14 +1,63 @@
+        self.expression, self.ordering = exprs[:ordering_exprs_i], exprs[ordering_exprs_i:]
+
+    def get_source_fields(self):
+        # do not return fields contributed by the ordering expressions, as
```
`# Filter out fields contributed....`
(chop comma before as)

timgraham commented on django/contrib/postgres/aggregates/general.py:
```diff
@@ -1,14 +1,63 @@
+    template = "%(function)s(%(expressions)s %(ordering)s)"
+
+    def __init__(self, expression, ordering=None, **extra):
+        if not isinstance(ordering, list) and ordering is not None:
```
`isinstance(ordering, (list, tuple))` (and add a test for the tuple case)
You can also borrow some documentation from ref/models/options.html (Options.ordering) -- the current proposed docs don't describe the string case or the fact that ordering can be a list/tuple.

timgraham commented on django/contrib/postgres/aggregates/general.py:
```diff
@@ -1,14 +1,63 @@
 
-class ArrayAgg(Aggregate):
+class OrderableAgg(Aggregate):
+    template = "%(function)s(%(expressions)s %(ordering)s)"
```
Use single quotes (unless a string contains a single quote) as described in [Python coding style](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/#python-style).

timgraham commented on django/contrib/postgres/aggregates/general.py:
```diff
@@ -1,14 +1,63 @@
 
 
-class ArrayAgg(Aggregate):
+class OrderableAgg(Aggregate):
```
Maybe this should be `OrderableAggMixin` and not inherit `Aggregate`? I'm thinking that other mixins could be added at some point.

florisdenhengst replied:
this is a good idea, will turn the ``OrderableAggregate`` into an ``OrderableAggMixin``

timgraham commented on django/contrib/postgres/aggregates/general.py:
```diff
@@ -1,14 +1,63 @@
+class OrderableAgg(Aggregate):
+    template = "%(function)s(%(expressions)s %(ordering)s)"
+
+    def __init__(self, expression, ordering=None, **extra):
```
`ordering=()` might simplify things.

timgraham commented on django/contrib/postgres/aggregates/general.py:
```diff
@@ -1,14 +1,63 @@
+        if not isinstance(ordering, list) and ordering is not None:
+            ordering = [ordering]
+        ordering = ordering or []
+        ordering = [(OrderBy(F(o[1:]), True) if isinstance(o, str) and o[0] == '-' else o) for o in ordering]
```
```
# Transform minus sign prefixed strings into an OrderBy() expression.
ordering = [
    (OrderBy(F(o[1:]), descending=True) if isinstance(o, str) and o[0] == '-' else o)
    for o in ordering
]
```

timgraham commented on django/contrib/postgres/aggregates/general.py:
```diff
@@ -1,14 +1,63 @@
+        return super().resolve_expression(*args, **kwargs)
+
+    def as_sql(self, compiler, connection):
+        # turn ordering parameter to ORDER BY sql clause
```
Turn (capitalize)
add period.
sql -> SQL

timgraham commented on django/contrib/postgres/aggregates/general.py:
```diff
@@ -1,14 +1,63 @@
+            for e in self.get_source_expressions()[:self._get_ordering_expressions_index()]]
+
+    def _get_ordering_expressions_index(self):
+        """
```
`"""Return the index at which the ordering expressions start."""`

timgraham commented on docs/ref/contrib/postgres/aggregates.txt:
```diff
@@ -22,16 +22,27 @@ General-purpose aggregation functions
+        unspecified.  When an ``ordering`` expression
+        (``F('some_field).asc()``) is provided, the elements in the returned
+        list are ordered by that expression.
+
```
Chop a blank line.

timgraham commented on docs/ref/contrib/postgres/aggregates.txt:
```diff
@@ -22,16 +22,27 @@ General-purpose aggregation functions
+.. class:: ArrayAgg(expression, distinct=False, ordering=None, **extra)
 
     Returns a list of values, including nulls, concatenated into an array.
+    Ordering within the array is unspecified by default.
```
I think this is redundant with the same sentence under the ordering attribute.

timgraham commented on docs/ref/contrib/postgres/aggregates.txt:
```diff
@@ -22,16 +22,27 @@ General-purpose aggregation functions
+
+        .. versionadded:: 2.0
+
+        By default the ordering of elements within the returned list is
```
By default,

timgraham commented on tests/postgres_tests/test_aggregates.py:
```diff
@@ -22,21 +22,54 @@ class TestGeneralAggregate(PostgreSQLTestCase):
+        )
+        for ordering, expected_output in ordering_test_cases:
+            with self.subTest(ordering=ordering, expected_output=expected_output):
+                values = AggregateTestModel.objects.aggregate(arrayagg=ArrayAgg('char_field',
```
Use hanging indent as described in [Python coding style](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/#python-style).
```
values = AggregateTestModel.objects.aggregate(
    arrayagg=ArrayAgg('char_field', ordering=ordering)
)
```

florisdenhengst commented:
@timgraham Thanks for the feedback, I processed your comments and squashed the commits.

timgraham commented on docs/ref/contrib/postgres/aggregates.txt:
```diff
@@ -22,7 +22,7 @@ General-purpose aggregation functions
 ------------
 
-.. class:: ArrayAgg(expression, distinct=False, **extra)
+.. class:: ArrayAgg(expression, distinct=False, ordering=None, **extra)
```
ordering=()

timgraham commented on django/contrib/postgres/aggregates/mixins.py:
```diff
@@ -0,0 +1,50 @@
+    def get_source_expressions(self):
+        return self.source_expressions + self.ordering
+
+#    def set_source_expressions(self, exprs):
```
I don't see any tests failing with this method removed. Is it required? If so, could you add a test to show why?

timgraham commented:
@florisdenhengst, do you have a chance to respond to those questions?

ngnpope commented:
@charettes Was looking at https://github.com/django/django/pull/7604#issuecomment-309862541 above where you mention only implementing orderable aggregates in `django.contrib.postgres`. In light of https://github.com/django/django/pull/9174, may your previous position on this change?

Obviously there are far fewer aggregates that support an order by clause than distinct, so perhaps this should be implemented as `supported_orderable_aggregates` to explicitly opt-in via the same mechanism in `check_expression_support`.

charettes commented:
@ngnpope until we've got at least one orderable aggregate supported on more than one backend ordering I still think that's it's better to focus on adding this only the `contrib.postgres`.

I think `GroupConcat` would be a good candidate for this. It would obsolete `contib.postgres.StringAgg` and would need to be marked as _unorderable_ on SQLite.

bobort commented:
I would really love to see this added to Django.  I just came across a use case for myself in which it would be incredibly useful.

bcaudell95 commented:
This feature would be similarly helpful for me.  Ordering an array aggregate is a nice feature in PostgresQL and would be appreciated in the framework.

vinayan3 commented:
Any updates on getting this merged?

carltongibson commented:
@vinayan3: It looks like there are outstanding issues from the previous reviews. It they can be addressed (either here or in a new PR) then we can uncheck "Patch needs improvement" [on the ticket](https://code.djangoproject.com/ticket/26067) and we can have another look.

bcaudell95 commented:
Is @florisdenhengst still around and working on this project?  I'd be happy to help if I could be of use, but I don't have write access to branches on his forked repo at the moment.

florisdenhengst commented:
I have to admit that I am far from an expert on ORM internals. Anything of value in this patch is thanks to the excellent reviewers. Specifically [this](https://github.com/django/django/pull/7604/commits/e91771813a07c0217106e73898f65ea9ce0c0ddc#r125748888) remark has halted me. Any help on this is much appreciated. I'm happy to create some more time to wrap this up coming days/weeks or permit access to my fork (I invited @bcaudell95 for starters).

carltongibson commented:
OK, thanks @florisdenhengst and @bcaudell95! Let me know (@mention me) if/when you run out of steam and I'll try and have a look too. 👍
